### PR TITLE
Api rtt

### PR DESCRIPTION
**Pull request by mattallinson**
Okay this should make the whole thing run from the API only but I think main.py might also require more updates, I've not tried to run a test verion of the whole thing yet, but made rtt.py generate lists of trains from searches and as far as i can tell from those it's working correctly.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -20,15 +20,23 @@
 AUTH_FILE = sys.argv[1]
 
 
-def make_twitter_api():
+def make_api_keys():
```
Comment (keirwl) [inline main.py]: I would have this be two separate funtions: one here for twitter, one in `realtimetrains.py`...

File: realtimetrains.py [hunk 2]
```diff
@@ -1,37 +1,24 @@
...
+TRAIN_SEARCH = "servicev2"
 DATE_FORMAT = "%Y/%m/%d"
 TIME_FORMAT = "%H%M"
-FRAC_TIMES = {
-    "¼": datetime.timedelta(seconds=15),
-    "½": datetime.timedelta(seconds=30),
-    "¾": datetime.timedelta(seconds=45)
-}
 ONE_DAY = datetime.timedelta(days=1)
 NO_SCHEDULE = "Couldn't find the schedule..."
```
Comment (keirwl) [inline realtimetrains.py:36]: ...then you can stick `rtt_api` here and not have to give it to every train.

File: realtimetrains.py [hunk 3]
```diff
@@ -136,7 +149,7 @@ def update_locations(self, soup):
 
     def populate(self):
         # print("Getting data for {}".format(self))
-        r = requests.get(self.url)
+        r = requests.get(self.url, auth = self.api_key)
```
Comment (keirwl) [inline realtimetrains.py]: No spaces around the = in kwargs. Same further down below.

File: realtimetrains.py [hunk 4]
```diff
@@ -145,90 +158,58 @@ def populate(self):
             return False
         else:
             self.webpage_checksum = md5sum
-
-        soup = BeautifulSoup(r.text, "html.parser")
-        if soup.text == NO_SCHEDULE:
+        
+        if 'realtimeActivated' not in r.json().keys():
```
Comment (keirwl) [inline realtimetrains.py]: `if blah not in dict` is the same as `if blah not in dict.keys()` and I think is a bit faster.

File: realtimetrains.py [hunk 5]
```diff
@@ -145,90 +158,58 @@ def populate(self):
...
             self.running = False
-            raise RuntimeError("schedule not found")
-        # Top of page shows schedule info, including if a
-        # runs-as-required train is active
-        schedule_info = soup.find("div",
-                                  attrs={"class": "detailed-schedule-info"})
-        # Text in schedule_info isn't tagged well
-        if "Running" in schedule_info.text:
+            #raise RuntimeError("schedule not found")
```
Comment (keirwl) [inline realtimetrains.py]: I think we can get rid of this. In the future we should add error checking around every `requests.get` call to see if there's a 404 or it's been a server error or something.

File: realtimetrains.py [hunk 6]
```diff
@@ -145,90 +158,58 @@ def populate(self):
...
-        return None
     # First four digits are in the simple form of HHMM
     loc_time = datetime.datetime.strptime(loc_timestring[:4],
                                           TIME_FORMAT).time()
     loc_datetime = datetime.datetime.combine(loc_date, loc_time)
-    # Sometimes there is a final fractional digit, whose value in
-    # seconds stored as timedeltas can be looked up
-    if len(loc_timestring) == 5:
-        loc_datetime += FRAC_TIMES[loc_timestring[4]]
+    # Sometimes the time is actual a 6 digit Hrs Mins Secs time
```
Comment (keirwl) [inline realtimetrains.py]: `actually`

File: realtimetrains.py [hunk 7]
```diff
@@ -145,90 +158,58 @@ def populate(self):
...
-                         "to", to_station, url_date, url_time])
+        search_url = "/".join([URL_PREFIX, LOCATION_SEARCH, station, "to", to_station, url_date])
     else:
-        return "/".join([URL_PREFIX, "search/advanced",
-                         station, url_date, url_time])
+        search_url = "/".join([URL_PREFIX, LOCATION_SEARCH, station, url_date])
+    
+    if to_time is not None: #adds time specific searching
+        time_string = to_time.strftime(TIME_FORMAT)
+        search_url += "/"+time_string
```
Comment (keirwl) [inline realtimetrains.py]: Spaces around the + operator.

File: realtimetrains.py [hunk 8]
```diff
@@ -145,90 +158,58 @@ def populate(self):
...
-
-    for row in rows:
-        id_url = row.find("a")["href"]
-        # url is in form of ".../train/H61429/2016/04/23/advanced"
-        # we want the UID, the third part of this
-        uid = id_url.split("/")[2]
-
-        trains.append(Train(uid, search_date))
+
+    url = _search_url(station, to_station = to_station, search_date = search_date,  to_time = time)
```
Comment (keirwl) [inline realtimetrains.py]: Spacing

**Review by keirwl — COMMENTED** [review]
Looking good. A few small comments, mainly about style. You've done something odd with the branching so let me fix that before merging (you'll need to do some stuff too that I'll walk you through).

File: README.md [hunk 9]
```diff
@@ -1,11 +1,25 @@
...
+a bot that auto-trainspots british nuclear waste trains 
+may be the most ballardian thing I've seen on this hell site
+```
+-- [@cszabla](https://twitter.com/cszabla/status/1072663099256254464?s=19)
+
+This is powered by RealTimeTrains http://www.realtimetrains.co.uk/
+API access for RTT is registered for at https://api.rtt.io/
+
+Thank you to Tom Cairns and the rest of the RTT for providing this service and for being speedy with tech support.
+Shut up and let them take your money https://secure.realtimetrains.co.uk/fundraising/
```
Comment (keirwl) [inline README.md:13]: I have given them some money!

File: realtimetrains.py (hunk 5 above)
Comment (mattallinson) [reply realtimetrains.py]: I'll use `r.raise_for_status` after the requests

File: realtimetrains.py (hunk 6 above)
Comment (mattallinson) [reply realtimetrains.py]: 🤓

File: main.py (hunk 1 above)
Comment (mattallinson) [reply main.py]: Done

**Comment by mattallinson** [comment]
Okay so I _think_ we might be nearly there with this. I've played around with RTT quite a bit now and it can generate a list of scheduled trains at a station, say which ones are running, update the times, correctly display the delay, say if the train is cancelled (which it couldn't before and might be useful for the twitter bot, not that I've ever noticed it tweet about a cancelled train before) and generally no longer have to rely on web-scraping which is soon to be killed. If you don't mind checking it one last time we can pull this into main and then get started with updating the twitter bot too.

File: main.py [hunk 10]
```diff
@@ -72,7 +72,7 @@ def make_tweets(train):
 def get_trains(routes):
     all_trains = []
     for route in routes:
-        trains = rtt.search(route["from"], to_station=route["to"])
+        trains = rtt.search(rtt_api, route["from"], to_station=route["to"])
```
Comment (keirwl) [inline main.py]: Excess space on the end of this line. Also it still refers to rtt_api.

File: main.py [hunk 11]
```diff
@@ -147,4 +147,4 @@ def main():
 
 
 if __name__ == '__main__':
-    main()
+    main()
```
Comment (keirwl) [inline main.py]: Needs a trailing newline.

File: realtimetrains.py [hunk 12]
```diff
@@ -1,37 +1,50 @@
...
-
 import datetime
 from hashlib import md5
-
-from bs4 import BeautifulSoup
 import requests
+import sys
+
+'''
+Creates the API key for realtime trains when imported when running a script
```
Comment (keirwl) [inline realtimetrains.py]: This string should be inside the fuction.

File: realtimetrains.py [hunk 13]
```diff
@@ -145,90 +189,64 @@ def populate(self):
...
-    rows = table.find_all("tr")[1:]
-
-    for row in rows:
-        id_url = row.find("a")["href"]
-        # url is in form of ".../train/H61429/2016/04/23/advanced"
-        # we want the UID, the third part of this
-        uid = id_url.split("/")[2]
 
+    url = _search_url(station, to_station=to_station, search_date=search_date,  to_time=time)
+    request = requests.get(url, auth = rtt_api)
```
Comment (keirwl) [inline realtimetrains.py]: Excess spaces around `auth=rtt_api`

**Review by keirwl — COMMENTED** [review]
Some style comments

**Pull request merged**